core/services/gateway/handlers/webapicapabilities/handler.go

```diff
@@ -1,6 +1,167 @@
+			ExecutionError: false,
+			StatusCode:     uint16(resp.StatusCode),
+			Headers:        resp.Headers,
+			Body:           resp.Body,
+		}
+	}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+	return &api.Message{
+		Body: api.MessageBody{
+			MessageId: msg.Body.MessageId,
+			Method:    msg.Body.Method,
+			DonId:     msg.Body.DonId,
+			Payload:   payloadBytes,
+		},
+	}, nil
+}
+
+func (h *handler) handleWebAPITargetMessage(ctx context.Context, msg *api.Message, nodeAddr string) error {
+	h.lggr.Debugw("handling web api target message", "messageId", msg.Body.MessageId, "nodeAddr", nodeAddr)
+	if !h.nodeRateLimiter.Allow(nodeAddr) {
+		return fmt.Errorf("rate limit exceeded for node %s", nodeAddr)
+	}
+	var targetPayload TargetRequestPayload
+	err := json.Unmarshal(msg.Body.Payload, &targetPayload)
+	if err != nil {
+		return err
+	}
+	// send message to target
+	timeout := time.Duration(targetPayload.TimeoutMs) * time.Millisecond
+	req := network.HTTPRequest{
+		Method:  targetPayload.Method,
+		URL:     targetPayload.URL,
+		Headers: targetPayload.Headers,
+		Body:    targetPayload.Body,
+		Timeout: timeout,
+	}
+	// this handle method must be non-blocking
+	// send response to node (target capability) async
+	// if there is a non-HTTP error (e.g. malformed request), send payload with success set to false and error messages
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		// not cancelled when parent is cancelled to ensure the goroutine can finish
+		newCtx := context.WithoutCancel(ctx)
+		newCtx, cancel := context.WithTimeout(newCtx, timeout)
+		defer cancel()
+		l := h.lggr.With("url", targetPayload.URL, "messageId", msg.Body.MessageId, "method", targetPayload.Method)
+		respMsg, err := h.sendHTTPMessageToClient(newCtx, req, msg)
+		if err != nil {
+			l.Errorw("error while sending HTTP request to external endpoint", "err", err)
+			payload := TargetResponsePayload{
+				ExecutionError: true,
+				ErrorMessage:   err.Error(),
+			}
+			payloadBytes, err2 := json.Marshal(payload)
+			if err2 != nil {
+				// should not happen
+				l.Errorw("error while marshalling payload", "err", err2)
+				return
+			}
+			respMsg = &api.Message{
+				Body: api.MessageBody{
+					MessageId: msg.Body.MessageId,
+					Method:    msg.Body.Method,
+					DonId:     msg.Body.DonId,
+					Payload:   payloadBytes,
+				},
+			}
+		}
+		err = h.don.SendToNode(newCtx, nodeAddr, respMsg)
+		if err != nil {
+			l.Errorw("failed to send to node", "err", err, "to", nodeAddr)
+			return
+		}
+	}()
+	return nil
+}
+
+func (h *handler) HandleNodeMessage(ctx context.Context, msg *api.Message, nodeAddr string) error {
+	switch msg.Body.Method {
+	case MethodWebAPITarget:
+		return h.handleWebAPITargetMessage(ctx, msg, nodeAddr)
+	default:
+		return fmt.Errorf("unsupported method: %s", msg.Body.Method)
+	}
+}
+
+func (h *handler) Start(context.Context) error {
+	return nil
+}
+
+func (h *handler) Close() error {
+	h.wg.Wait()
```

Comment:
Hmmm... do we need a way to signal to ongoing requests to close here? If their timeout is quite long, then waiting for them to expire could mean we wait for a long time.

Happy for this to be a follow up though

Reply:
I am not sure if it's completely necessary since timeout will be in the order of seconds for typical HTTP requests. This reminds me though we should validate the timeout on the capability side to have a reasonable max value.